engines/e_pkcs11.c

```diff
@@ -0,0 +1,841 @@
+            goto err;
+        }
+    }
+
+ err:
+    OPENSSL_free(strbuf);
+#endif
+
+    return 0;
+}
+
+static int pkcs11_parse(PKCS11_CTX *ctx, const char *path)
+{
+    char *id, *module_path = NULL;
+    char *pin = NULL, *label = NULL;
+    char *slotid = OPENSSL_malloc(2);
+
+    if (slotid == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    if (strncmp(path, "pkcs11:", 7) == 0) {
+        path += 7;
+	if (ctx->module_path == NULL &&
+            !pkcs11_parse_uri(path,"module-path=", &module_path))
+            goto err;
+        if (ctx->module_path == NULL) ctx->module_path = module_path;
+	if (!pkcs11_parse_uri(path,"id=", &id) &&
+            !pkcs11_parse_uri(path,"object=", &label)) {
+            PKCS11_trace("ID and OBJECT are null\n");
+            goto err;
+        }
+	if (!pkcs11_parse_uri(path,"slot-id=", &slotid)) {
+           slotid[0] = '0';
+        }
+        pkcs11_parse_uri(path,"pin-value=", &pin);
+        if (pin != NULL) ctx->pin = (CK_BYTE *) pin;
+    } else if (path == NULL) {
+       PKCS11_trace("inkey is null\n");
+       goto err;
+    } else {
+        if (ctx->module_path == NULL) {
+            PKCS11_trace("Module path is null\n");
+            goto err;
+        }
+        id = OPENSSL_strdup(path);
+
+        if (id == NULL) {
+            PKCS11err(PKCS11_F_PKCS11_PARSE, ERR_R_MALLOC_FAILURE);
+            goto err;
+        }
+
+        slotid[0] = '0';
+    }
+    if (label != NULL)
+        ctx->label = (CK_BYTE *) label;
+    else
+        ctx->id = (CK_BYTE *) id;
+    ctx->slotid = (CK_SLOT_ID) atoi(slotid);
+
+    if (ctx->pin == NULL) {
+        if (!pkcs11_get_console_pin(&pin))
+            goto err;
+        ctx->pin = (CK_BYTE *) pin;
+        if (ctx->pin == NULL) {
+            PKCS11_trace("PIN is invalid\n");
+            goto err;
+        }
+    }
+
+    return 1;
+
+ err:
+    return 0;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data)
+{
+    CK_ULONG kt, key_class;
+    CK_ATTRIBUTE key_type[2];
+    CK_RV rv;
+    PKCS11_CTX *ctx;
+
+    key_type[0].type = CKA_CLASS;
+    key_type[0].pValue = &key_class;
+    key_type[0].ulValueLen = sizeof(key_class);
+    key_type[1].type = CKA_KEY_TYPE;
+    key_type[1].pValue = &kt;
+    key_type[1].ulValueLen = sizeof(kt);
+    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
+
+    if (!pkcs11_parse(ctx, path))
+        goto err;
+
+    rv = pkcs11_initialize(ctx->module_path);
+    if (rv != CKR_OK)
+        goto err;
+    if (!pkcs11_get_slot(ctx))
+        goto err;
+    if (!pkcs11_start_session(ctx))
+        goto err;
+    if (!pkcs11_login(ctx, CKU_USER))
+        goto err;
+    if (!pkcs11_get_private_key(ctx))
+        goto err;
+
+    rv = pkcs11_funcs->C_GetAttributeValue(ctx->session, ctx->key,
```

Comment:
this `pkcs11_funcs->C_GetAttributeValue` call should be a separate methods, or moved into the previous one.